src/WebJobs.Script.WebHost/ContainerManagement/LinuxContainerActivityPublisher.cs

```diff
@@ -169,81 +192,72 @@ private async Task FlushFunctionExecutionActivities()
             }
-            catch (ObjectDisposedException)
+
+            try
             {
-                // might race with dispose
+                _uniqueActivities.Add(activity);
             }
-            catch (Exception e)
+            finally
             {
-                _logger.LogError(e, nameof(SetTimerInterval));
+                _activitiesLock.ExitWriteLock();
             }
+
+            return true;
         }
 
-        public void Dispose()
+        private bool TryGetCurrentActivities(IList<ContainerFunctionExecutionActivity> currentActivities)
         {
-            if (_timer != null)
+            if (!_activitiesLock.TryEnterWriteLock(LockTimeOutMs))
             {
-                _timer?.Dispose();
-                _timer = null;
+                return false;
             }
-        }
 
-        private bool PublishActivity(ContainerFunctionExecutionActivity activity)
-        {
-            if (_activitiesLock.TryEnterWriteLock(LockTimeOutMs))
+            try
             {
-                try
-                {
-                    _uniqueActivities.Add(activity);
-                }
-                finally
+                foreach (var activity in _uniqueActivities)
                 {
-                    _activitiesLock.ExitWriteLock();
+                    currentActivities.Add(activity);
                 }
-                return true;
-            }
 
-            return false;
+                _uniqueActivities.Clear();
+            }
+            finally
+            {
+                _activitiesLock.ExitWriteLock();
+            }
+            return true;
         }
 
-        private bool TryGetCurrentActivities(IList<ContainerFunctionExecutionActivity> currentActivities)
+        public void PublishFunctionExecutionActivity(ContainerFunctionExecutionActivity activity)
         {
-            if (_activitiesLock.TryEnterWriteLock(LockTimeOutMs))
+            if (_standbyOptions.CurrentValue.InStandbyMode)
             {
-                try
-                {
-                    foreach (var activity in _uniqueActivities)
-                    {
-                        currentActivities.Add(activity);
-                    }
-                    _uniqueActivities.Clear();
-                }
-                finally
-                {
-                    _activitiesLock.ExitWriteLock();
-                }
-                return true;
+                return;
             }
 
-            return false;
+            if (!PublishActivity(activity))
+            {
+                _logger.LogWarning($"Failed to add activity {activity}");
+            }
         }
 
-        public void PublishFunctionExecutionActivity(ContainerFunctionExecutionActivity activity)
+        public async ValueTask DisposeAsync()
         {
-            if (!_standbyOptions.CurrentValue.InStandbyMode)
+            // Wait for the publishing task to complete
+            if (_publishingTask != null)
```

Comment:
Do you need to cancel `_publishCts` before `await _publishingTask`? Otherwise if this is disposed without stopping, this would never complete.

Reply:
Cancellation is expected in `StopAsync`, but I'll add some logic to handle a potential disposal where cancellation was not called.